server/src/main/java/org/elasticsearch/search/lookup/SearchLookup.java

```diff
@@ -50,39 +50,46 @@ public class SearchLookup implements SourceProvider {
         Supplier<SearchLookup>,
         MappedFieldType.FielddataOperation,
         IndexFieldData<?>> fieldDataLookup;
+    private final TriFunction<MappedFieldType, Supplier<SearchLookup>, MappedFieldType.FielddataOperation, Boolean> fieldDataPredicate;
+
     private final Function<LeafReaderContext, LeafFieldLookupProvider> fieldLookupProvider;
 
     /**
      * Create a new SearchLookup, using the default stored fields provider
-     * @param fieldTypeLookup   defines how to look up field types
-     * @param fieldDataLookup   defines how to look up field data
-     * @param sourceProvider    defines how to look up the source
+     * @param fieldTypeLookup    defines how to look up field types
+     * @param fieldDataLookup    defines how to look up field data
+     * @param fieldDataPredicate defines if it is possible to access field data
+     * @param sourceProvider     defines how to look up the source
      */
     public SearchLookup(
         Function<String, MappedFieldType> fieldTypeLookup,
         TriFunction<MappedFieldType, Supplier<SearchLookup>, MappedFieldType.FielddataOperation, IndexFieldData<?>> fieldDataLookup,
+        TriFunction<MappedFieldType, Supplier<SearchLookup>, MappedFieldType.FielddataOperation, Boolean> fieldDataPredicate,
```

Comment:
Here I introduced a second `TriFunction` instead of an interface that includes both the lookup and the predicate - see `SearchExecutionContext` for that approach.
I'd say we keep one for consistency, but it's not clear to me which one is more readable so I kept both and I am asking for feedback.